File: src/plan.rs
use crate::model::{Language, Mode, SessionRecord, TokenKind};
use chrono::{Duration, Utc};
use std::collections::BTreeMap;

#[derive(Debug, Clone)]
pub struct PracticePlan {
    pub focus_words: Vec<String>,
    pub focus_symbols: Vec<String>,
    pub focus_code: Vec<String>,
    pub advice: Vec<String>,
    pub recommended_mode: Mode,
}

#[derive(Debug, Default)]
struct Aggregate {
    occurrences: u32,
    errors: u32,
    delay_sum: u64,
    duration_sum: u64,
}

impl Aggregate {
    fn add(&mut self, delay: u64, duration: u64, errors: u32) {
        self.occurrences += 1;
        self.errors += errors;
        self.delay_sum += delay;
        self.duration_sum += duration;
    }

    fn avg_delay(&self) -> f64 {
        if self.occurrences == 0 {
            return 0.0;
        }
        self.delay_sum as f64 / self.occurrences as f64
    }

    fn avg_duration(&self) -> f64 {
        if self.occurrences == 0 {
            return 0.0;
        }
        self.duration_sum as f64 / self.occurrences as f64
    }

    fn score(&self) -> f64 {
        self.avg_delay() + self.avg_duration() * 0.25 + f64::from(self.errors) * 300.0
    }
}

pub fn build_plan(records: &[SessionRecord], language: Language) -> PracticePlan {
    let recent_cutoff = Utc::now() - Duration::days(21);
    let recent: Vec<&SessionRecord> = records
        .iter()
        .filter(|record| record.started_at >= recent_cutoff)
        .collect();

    if recent.is_empty() {
        return PracticePlan {
            focus_words: vec![
                "return".into(),
                "function".into(),
                "current".into(),
                "response".into(),
                "useEffect".into(),
            ],
            focus_symbols: vec![
                "=>".into(),
                "!==".into(),
                "&&".into(),
                "_".into(),
                "{}".into(),
            ],
            focus_code: vec!["useState".into(), "items.map".into(), "!== null".into()],
            advice: no_history_advice(language),
            recommended_mode: Mode::Chars,
        };
    }

    let mut words = BTreeMap::<String, Aggregate>::new();
    let mut symbols = BTreeMap::<String, Aggregate>::new();
    let mut code_terms = BTreeMap::<String, Aggregate>::new();
    let mut total_typed = 0usize;
    let mut total_key_correct = 0usize;
    let mut total_backspaces = 0u32;

    for record in &recent {
        let typed_len = effective_typed_len(record);
        total_typed += typed_len;
        total_key_correct +=
            (record.accuracy.clamp(0.0, 100.0) / 100.0 * typed_len as f64).round() as usize;
        total_backspaces += record.backspace_count;

        if record.token_stats.is_empty() {
            for (token, errors) in &record.error_tokens {
                let aggregate = if is_word_like_token(token) {
                    words.entry(token.clone()).or_default()
                } else {
                    symbols.entry(token.clone()).or_default()
                };
                aggregate.add(0, 0, *errors);
            }
        } else {
            for stat in &record.token_stats {
                let aggregate = match stat.kind {
                    TokenKind::Word => words.entry(stat.token.clone()).or_default(),
                    TokenKind::Symbol => symbols.entry(stat.token.clone()).or_default(),
                    TokenKind::Code => code_terms.entry(stat.token.clone()).or_default(),
                };
                aggregate.add(stat.start_delay_ms, stat.duration_ms, stat.errors);
            }
        }
    }

    let focus_words = top_keys(&words, 16)
        .into_iter()
        .filter(|word| is_focus_word(word))
        .take(6)
        .collect::<Vec<_>>();
    let focus_symbols = top_keys(&symbols, 6);
    let mut focus_code = top_keys(&code_terms, 8)
        .into_iter()
        .filter(|term| term.chars().count() >= 2)
        .take(4)
        .collect::<Vec<_>>();
    for word in focus_words.iter().take(3) {
        if !focus_code.contains(word) {
            focus_code.push(word.clone());
        }
    }

    let accuracy = if total_typed == 0 {
        0.0
    } else {
        total_key_correct as f64 / total_typed as f64 * 100.0
    };

    let mut advice = Vec::new();
    if accuracy < 95.0 {
        advice.push(match language {
            Language::Zh => "正确率低于 95%。下一轮缩短一点，慢一点打准。".into(),
            Language::En => {
                "Accuracy is below 95%. Keep the next session shorter and type deliberately.".into()
            }
        });
    }
    if total_backspaces > recent.len() as u32 * 12 {
        advice.push(match language {
            Language::Zh => "退格偏多。放慢一点，避免反复修正。".into(),
            Language::En => {
                "Backspace count is high. Slow down slightly and avoid correction loops.".into()
            }
        });
    }
    if !focus_words.is_empty() {
        advice.push(match language {
            Language::Zh => format!("复盘单词和标识符：{}。", focus_words.join(", ")),
            Language::En => format!("Review words and identifiers: {}.", focus_words.join(", ")),
        });
    }
    if !focus_symbols.is_empty() {
        advice.push(match language {
            Language::Zh => format!("复盘代码符号：{}。", focus_symbols.join(", ")),
            Language::En => format!("Review code symbols: {}.", focus_symbols.join(", ")),
        });
    }
    if advice.is_empty() {
        advice.push(match language {
            Language::Zh => "表现比较稳定。可以用混合模式，并加入更多真实代码片段。".into(),
            Language::En => {
                "Performance looks stable. Use mixed mode and include more real code snippets."
                    .into()
            }
        });
    }

    let symbol_pressure = symbols.values().map(Aggregate::score).sum::<f64>();
    let word_pressure = words.values().map(Aggregate::score).sum::<f64>();
    let recommended_mode = if symbol_pressure > word_pressure * 1.15 {
        Mode::Symbols
    } else if word_pressure > symbol_pressure * 1.15 {
        Mode::Words
    } else {
        Mode::Mixed
    };

    PracticePlan {
        focus_words,
        focus_symbols,
        focus_code,
        advice,
        recommended_mode,
    }
}

fn no_history_advice(language: Language) -> Vec<String> {
    match language {
        Language::Zh => vec![
            "还没有练习记录。先从基础字符开始，建立一条基准数据。".into(),
            "先把正确率打稳，再追速度，这样后续分析才有用。".into(),
        ],
        Language::En => vec![
            "No history yet. Start with a mixed baseline session.".into(),
            "Focus on accuracy before speed so the data is useful.".into(),
        ],
    }
}

fn top_keys(map: &BTreeMap<String, Aggregate>, limit: usize) -> Vec<String> {
    let mut scored: Vec<(&String, &Aggregate)> = map.iter().collect();
    scored.sort_by(|(_, a), (_, b)| {
        b.score()
            .partial_cmp(&a.score())
            .unwrap_or(std::cmp::Ordering::Equal)
    });

    scored
        .into_iter()
        .filter(|(_, aggregate)| aggregate.occurrences > 0)
        .take(limit)
        .map(|(key, _)| key.clone())
        .collect()
}

fn is_focus_word(word: &str) -> bool {
    let len = word.chars().count();
    if len < 3 {
        return false;
    }
    word.chars().any(|ch| ch.is_ascii_uppercase()) || len >= 5
}

fn is_word_like_token(token: &str) -> bool {
    token.chars().any(|ch| ch.is_ascii_alphabetic())
        && token
            .chars()
            .all(|ch| ch.is_ascii_alphanumeric() || ch == '_')
}

fn effective_typed_len(record: &SessionRecord) -> usize {
    if record.typed_len > 0 {
        return record.typed_len;
    }
    record.user_input.chars().count().max(record.correct_chars)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn build_plan_uses_legacy_error_tokens() {
        let mut record = SessionRecord {
            started_at: Utc::now(),
            typed_len: 10,
            accuracy: 80.0,
            ..SessionRecord::default()
        };
        record.error_tokens.insert("response".to_string(), 3);
        record.error_tokens.insert("=>".to_string(), 2);

        let plan = build_plan(&[record], Language::Zh);

        assert!(plan.focus_words.contains(&"response".to_string()));
        assert!(plan.focus_symbols.contains(&"=>".to_string()));
    }

    #[test]
    fn build_plan_uses_legacy_typed_len_fallback() {
        let record = SessionRecord {
            started_at: Utc::now(),
            typed_len: 0,
            correct_chars: 20,
            accuracy: 100.0,
            user_input: "abcdefghijklmnopqrst".to_string(),
            ..SessionRecord::default()
        };

        let plan = build_plan(&[record], Language::Zh);

        assert!(
            !plan
                .advice
                .iter()
                .any(|item| item.contains("正确率低于 95%"))
        );
    }
}

File: src/training/plan.ts
import type {
  KeyEventRecord,
  Language,
  Mode,
  PracticePlan,
  SessionRecord,
  TokenKind,
} from "../domain/model";
import { isNumberedTemplateIdentifier } from "./generatedIdentifier";

export const PLAN_HISTORY_DAYS = 21;

interface Aggregate {
  occurrences: number;
  errors: number;
  delaySum: number;
  durationSum: number;
}

export function buildPlan(
  records: SessionRecord[],
  language: Language,
  now: Date = new Date(),
): PracticePlan {
  const recentCutoffMs = now.getTime() - PLAN_HISTORY_DAYS * 24 * 60 * 60 * 1000;
  const recent = records.filter((record) => {
    const startedAtMs = Date.parse(record.started_at);
    return Number.isFinite(startedAtMs) && startedAtMs >= recentCutoffMs;
  });

  if (recent.length === 0) {
    return {
      focus_words: ["return", "function", "current", "response", "useEffect"],
      focus_symbols: ["=>", "!==", "&&", "_", "{}"],
      focus_code: ["useState", "items.map", "!== null"],
      focus_keys: [],
      advice: noHistoryAdvice(language),
      recommended_mode: "chars",
      has_recent_history: false,
    };
  }

  const words = new Map<string, Aggregate>();
  const symbols = new Map<string, Aggregate>();
  const codeTerms = new Map<string, Aggregate>();
  const keys = new Map<string, Aggregate>();
  let totalTyped = 0;
  let totalKeyCorrect = 0;
  let totalBackspaces = 0;

  for (const record of recent) {
    const typedLen = effectiveTypedLen(record);
    totalTyped += typedLen;
    totalKeyCorrect += Math.round((clamp(record.accuracy, 0, 100) / 100) * typedLen);
    totalBackspaces += record.backspace_count;

    if (record.token_stats.length === 0) {
      for (const [token, errors] of sortedRecordEntries(record.error_tokens)) {
        if (isNumberedTemplateIdentifier(token)) {
          continue;
        }
        addAggregate(isWordLikeToken(token) ? words : symbols, token, 0, 0, errors);
      }
    } else {
      for (const stat of record.token_stats) {
        if (isNumberedTemplateIdentifier(stat.token)) {
          continue;
        }
        addAggregate(
          aggregateMapForTokenKind(stat.kind, words, symbols, codeTerms),
          stat.token,
          stat.start_delay_ms,
          stat.duration_ms,
          stat.errors,
        );
      }
    }

    for (const [key, count] of recordKeyErrors(record)) {
      addAggregate(keys, key, 0, 0, count);
    }
  }

  const focusWords = topKeys(words, 16).filter(isFocusWord).slice(0, 6);
  const focusSymbols = topKeys(symbols, 6);
  const focusCode = topKeys(codeTerms, 8)
    .filter((term) => Array.from(term).length >= 2)
    .slice(0, 4);
  for (const word of focusWords.slice(0, 3)) {
    if (!focusCode.includes(word)) {
      focusCode.push(word);
    }
  }
  const focusKeys = topKeys(keys, 8);

  const accuracy = totalTyped === 0 ? 0 : (totalKeyCorrect / totalTyped) * 100;
  const advice: string[] = [];

  if (accuracy < 95) {
    advice.push(
      language === "zh"
        ? "正确率低于 95%。下一轮缩短一点，慢一点打准。"
        : "Accuracy is below 95%. Keep the next session shorter and type deliberately.",
    );
  }
  if (totalBackspaces > recent.length * 12) {
    advice.push(
      language === "zh"
        ? "退格偏多。放慢一点，避免反复修正。"
        : "Backspace count is high. Slow down slightly and avoid correction loops.",
    );
  }
  if (focusWords.length > 0) {
    advice.push(
      language === "zh"
        ? `复盘单词和标识符：${focusWords.join(", ")}。`
        : `Review words and identifiers: ${focusWords.join(", ")}.`,
    );
  }
  if (focusSymbols.length > 0) {
    advice.push(
      language === "zh"
        ? `复盘代码符号：${focusSymbols.join(", ")}。`
        : `Review code symbols: ${focusSymbols.join(", ")}.`,
    );
  }
  if (focusKeys.length > 0) {
    advice.push(
      language === "zh"
        ? `补强键位热区：${focusKeys.join(", ")}。`
        : `Reinforce key hot spots: ${focusKeys.join(", ")}.`,
    );
  }
  if (advice.length === 0) {
    advice.push(
      language === "zh"
        ? "表现比较稳定。可以用混合模式，并加入更多真实代码片段。"
        : "Performance looks stable. Use mixed mode and include more real code snippets.",
    );
  }

  const symbolPressure = sumScores(symbols);
  const wordPressure = sumScores(words);
  const recommendedMode: Mode =
    symbolPressure > wordPressure * 1.15
      ? "symbols"
      : wordPressure > symbolPressure * 1.15
        ? "words"
        : "mixed";

  return {
    focus_words: focusWords,
    focus_symbols: focusSymbols,
    focus_code: focusCode,
    focus_keys: focusKeys,
    advice,
    recommended_mode: recommendedMode,
    has_recent_history: true,
  };
}

function aggregateMapForTokenKind(
  kind: TokenKind,
  words: Map<string, Aggregate>,
  symbols: Map<string, Aggregate>,
  codeTerms: Map<string, Aggregate>,
): Map<string, Aggregate> {
  switch (kind) {
    case "word":
      return words;
    case "symbol":
      return symbols;
    case "code":
      return codeTerms;
  }
}

function addAggregate(
  map: Map<string, Aggregate>,
  key: string,
  delay: number,
  duration: number,
  errors: number,
): void {
  const aggregate = map.get(key) ?? {
    occurrences: 0,
    errors: 0,
    delaySum: 0,
    durationSum: 0,
  };
  aggregate.occurrences += 1;
  aggregate.errors += errors;
  aggregate.delaySum += delay;
  aggregate.durationSum += duration;
  map.set(key, aggregate);
}

function recordKeyErrors(record: SessionRecord): Array<[string, number]> {
  const counts = new Map<string, number>();
  for (const event of record.key_events) {
    if (event.action === "insert" && !event.correct) {
      const key = event.expected ?? event.input;
      const label = key === null ? "extra" : keyBucketForChar(key);
      counts.set(label, (counts.get(label) ?? 0) + 1);
    }
  }

  if (record.key_events.length === 0) {
    for (const [label, count] of sortedRecordEntries(record.error_chars)) {
      const bucket = keyBucketForLabel(label);
      counts.set(bucket, (counts.get(bucket) ?? 0) + count);
    }
  }

  return [...counts.entries()].sort(([left], [right]) => left.localeCompare(right));
}

function noHistoryAdvice(language: Language): string[] {
  if (language === "zh") {
    return [
      "还没有练习记录。先从基础字符开始，建立一条基准数据。",
      "先把正确率打稳，再追速度，这样后续分析才有用。",
    ];
  }
  return [
    "No history yet. Start with a mixed baseline session.",
    "Focus on accuracy before speed so the data is useful.",
  ];
}

function topKeys(map: Map<string, Aggregate>, limit: number): string[] {
  return [...map.entries()]
    .sort(([leftKey, left], [rightKey, right]) => {
      const scoreDelta = aggregateScore(right) - aggregateScore(left);
      return scoreDelta === 0 ? leftKey.localeCompare(rightKey) : scoreDelta;
    })
    .filter(([, aggregate]) => aggregate.occurrences > 0)
    .slice(0, limit)
    .map(([key]) => key);
}

function isFocusWord(word: string): boolean {
  const length = Array.from(word).length;
  if (length < 3) {
    return false;
  }
  return /[A-Z]/u.test(word) || length >= 5;
}

function isWordLikeToken(token: string): boolean {
  return /[A-Za-z]/u.test(token) && /^[A-Za-z0-9_]+$/u.test(token);
}

function keyBucketForLabel(label: string): string {
  switch (label) {
    case "<space>":
      return "space";
    case "\\n":
      return "enter";
    case "\\t":
      return "tab";
    default: {
      const first = Array.from(label)[0];
      return first === undefined ? "" : keyBucketForChar(first);
    }
  }
}

function keyBucketForChar(value: string): string {
  switch (value) {
    case "!":
    case "1":
      return "1";
    case "@":
    case "2":
      return "2";
    case "#":
    case "3":
      return "3";
    case "$":
    case "4":
      return "4";
    case "%":
    case "5":
      return "5";
    case "^":
    case "6":
      return "6";
    case "&":
    case "7":
      return "7";
    case "*":
    case "8":
      return "8";
    case "(":
    case "9":
      return "9";
    case ")":
    case "0":
      return "0";
    case "_":
    case "-":
      return "-";
    case "+":
    case "=":
      return "=";
    case "~":
    case "`":
      return "`";
    case "{":
    case "[":
      return "[";
    case "}":
    case "]":
      return "]";
    case "|":
    case "\\":
      return "\\";
    case ":":
    case ";":
      return ";";
    case '"':
    case "'":
      return "'";
    case "<":
    case ",":
      return ",";
    case ">":
    case ".":
      return ".";
    case "?":
    case "/":
      return "/";
    case " ":
      return "space";
    case "\n":
      return "enter";
    case "\t":
      return "tab";
    default:
      return /^[A-Za-z]$/u.test(value) ? value.toLowerCase() : value;
  }
}

function effectiveTypedLen(record: SessionRecord): number {
  if (record.typed_len > 0) {
    return record.typed_len;
  }
  return Math.max(Array.from(record.user_input).length, record.correct_chars);
}

function sortedRecordEntries(record: Record<string, number>): Array<[string, number]> {
  return Object.entries(record).sort(([left], [right]) => left.localeCompare(right));
}

function sumScores(map: Map<string, Aggregate>): number {
  return [...map.values()].reduce((sum, aggregate) => sum + aggregateScore(aggregate), 0);
}

function aggregateScore(aggregate: Aggregate): number {
  return (
    average(aggregate.delaySum, aggregate.occurrences) +
    average(aggregate.durationSum, aggregate.occurrences) * 0.25 +
    aggregate.errors * 300
  );
}

function average(sum: number, count: number): number {
  return count === 0 ? 0 : sum / count;
}

function clamp(value: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, value));
}

function assertNever(value: never): never {
  throw new Error(`Unexpected token kind: ${value}`);
}
